stan/math/prim/err/check_lower_triangular.hpp

```diff
@@ -15,27 +15,27 @@ namespace math {
  * A matrix x is not lower triangular if there is a non-zero entry
  * x[m, n] with m &lt; n. This function only inspects the upper
  * triangular portion of the matrix, not including the diagonal.
- * @tparam T Type of scalar of the matrix
+ * @tparam T Type of the matrix
  * @param function Function name (for error messages)
  * @param name Variable name (for error messages)
  * @param y Matrix to test
  * @throw <code>std::domain_error</code> if the matrix is not
  *   lower triangular or if any element in the upper triangular
  *   portion is NaN
  */
-template <typename T_y>
-inline void check_lower_triangular(
-    const char* function, const char* name,
-    const Eigen::Matrix<T_y, Eigen::Dynamic, Eigen::Dynamic>& y) {
+template <typename T_y, require_eigen_t<T_y>* = nullptr>
+inline void check_lower_triangular(const char* function, const char* name,
+                                   const T_y& y) {
+  const Eigen::Ref<const plain_type_t<T_y>>& y_ref = y;
   for (int n = 1; n < y.cols(); ++n) {
     for (int m = 0; m < n && m < y.rows(); ++m) {
-      if (y(m, n) != 0) {
+      if (y_ref(m, n) != 0) {
```

Comment:
Optional:

Performance wise, do you think it's worth replacing this loop with something like

```cpp
if((A.template triangularView<Eigen::StrictlyUpper>().nestedExpression().array() == 0).any()) {
/// ...
}
```

https://godbolt.org/z/u2oMaB

Reply:
Yes, but if we do it that way we lose information at which index the error is detected. So if we detect the error we than have to repeat what the function does now to find the index. 

I am not sure if it is worth complicating the code unless we find a use case where this function is something close to a bottleneck. I think it is mostly used next to matrix decompositions, which are an order of magnitude slower than this.

Reply:
Cool cool yeah I think having that info is important